README.md
# About

**Pooling thrift connection with zookeeper in distributed environment.**

A module that integrates zookeeper with thrift interfaces with connection pooling to make them more robust.
For pooling different server instances of a service, it uses [Node pool cluster](https://www.npmjs.com/package/node-pool-cluster).
[Node zookeeper client](https://www.npmjs.com/package/node-zookeeper-client) is used to connect to zookeeper and 
[Thrift](https://www.npmjs.com/package/thrift) for creating thrift connections.

## Table of contents

* [Installation](#installation)
* [Usage](#usage)
* [Config Option](#config-option)
* [History](#history)
* [License](#license) (The MIT License)

## Installation

```
$ npm install thrift-zookeeper-client
```

## Usage

```javascript
var thrift = require('thrift'),
  myServiceClient = require('./gen-nodejs/Service'),
  thriftZkClient = require('thrift-zookeeper-client'),
  zkProps = {
      zkUrl:"localhost:2181",
      zkPath: "/com/test/services/qa"
  };
  
  thriftZkClient.getService(zkProps, "myService", myServiceClient, function(thriftMyServiceClient){
        //callback      
  });
  
  //to make TFramedTransport 
  thriftZkClient.getService(zkProps, "myService", myServiceClient, function(thriftMyServiceClient){
        //callback      
  }, true);
  
```

## Config Option

You can control some of the pooling options, similar to generic node pool.

```
  var thrift = require('thrift'),
    myServiceClient = require('./gen-nodejs/Service'),
    thriftZkClient = require('thrift-zookeeper-client'),
    zkProps = {
        zkUrl:"localhost:2181",
        zkPath: "/com/test/services/qa"
    };
    
    thriftZkClient.config({
        max: 5 //max connections in a pool, default is 3,
        min: 1 //min connections in a pool, default is 0,
        idleTimeoutMillis: 15000 //max milliseconds a resource can go unused before it should be destroyed (default 20000)
        refreshIdle: true, //boolean that specifies whether idle resources at or below the min threshold
                           //should be destroyed/re-created.  optional (default=true)
        reapIntervalMillis: 2000 //frequency to check for idle resources (default 1000),
    });
    
    thriftZkClient.getService(zkProps, "myService", myServiceClient, function(thriftMyServiceClient){
          //callback      
    });
```


## History

The history has been moved to the [CHANGELOG](ChangeLog.md)

## License

The MIT License (MIT)

Copyright (c) 2016 Kapil Joshi  <<kjoshi1988@gmail.com>>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

ChangeLog.md
## [0.1.2] - Feb 19, 2016
* Added support for TFramedTransport
* Minor bug fixed, for watcher.

## [0.1.1] - Feb 8, 2016
* Documentation and Config option support added
* Adding logic to fetch resource from another pool, when current pool throws error while acquiring resource.
* License added

## [0.1.0] - Feb 4, 2016

* First release.

lib/helper/zkHelper.js
var zookeeper = require('node-zookeeper-client');
var async = require('async');
var serviceClientWrapper = require("./serviceClientWrapper");
var _util = require('./util');
var _initCbQueue = {}, _clusterMap = {}, _serviceClients = {}
    , _zooKeeperClient, _config = {};


var zkHelper = {};

/**
 * Sets config options, like min/max/timeout of resources in a pool.
 * Defaults opts are:
 * <pre>
 *     max: 3
 *     min: 0
 *     idleTimeoutMillis: 20000(ms)
 * </pre>
 *
 */
zkHelper.config = function (opts) {
    _config.min = opts.min || 3;
    _config.max = opts.max || 0;
    _config.idleTimeoutMillis = opts.idleTimeoutMillis || 20000;
    _config.refreshIdle = opts.refreshIdle || true;
    _config.reapIntervalMillis = opts.reapIntervalMillis || 1000;
};

/**
 * This method creates thrift service client. It first checks, whether a client is already in cache or not.
 * If service client is already there in cache, then the callback is called with service client as argument,
 * otherwise it creates a new client.
 *
 *
 * @param {Object} zkProps, Object containing zooKeeper server ip, port and path
 * @param {String} serviceName, Name of the service to be fetched
 * @param {Object} serviceClient, thrift client for the service
 * @param {Function} callBack, this will be called, after creation of thrift service client.
 * @param {async} async, to enable TFramed protocol
 */
zkHelper.getService = function (zkProps, serviceName, serviceClient, callBack, async) {
    var _serviceClient = _serviceClients[_util.getServiceClientCacheKey(zkProps, serviceName)];
    if (!!_serviceClient) {
        callBack(_serviceClient);
    } else {
        zkHelper.createServiceClient(zkProps, serviceName, serviceClient, async, callBack);
    }
};


/**
 * Fetches service details hosted at the specified path.
 * It also adds watcher on the node, which will be triggered when service node is either deleted, modified or more
 * children are added.
 * Until the children details are fetched and client is created, all the request(for the same service client) callbacks
 * are queued.
 *
 * @param {Object} zkProps, Object containing zooKeeper server ip, port and path
 * @param {String} serviceName, Name of the service to be fetched
 * @param {Object} serviceClient, thrift client for the service
 * @param {async} async, to enable TFramed protocol
 * @param {Function} callBack
 */
zkHelper.createServiceClient = function (zkProps, serviceName, serviceClient, async, callBack) {
    var _cacheKey = _util.getServiceClientCacheKey(zkProps, serviceName);
    var _cbQueue = _initCbQueue[_cacheKey];
    if (!!_cbQueue) {
        !!callBack && _cbQueue.push(callBack);
    } else {
        _cbQueue = [];
        _initCbQueue[_cacheKey] = _cbQueue;
        !!callBack && _cbQueue.push(callBack);
        callZookeeper(zkProps.zkUrl, function (zkClient) {
            zkClient.getChildren(zkProps.zkPath + "/nodeStatus/" + serviceName,
                zkWatcher.bind(zkClient, zkProps, serviceName, serviceClient, async),
                zkProcessChildren.bind(zkClient, zkProps, serviceName, serviceClient, async)
            );
        });
    }
};


/**
 * Creates zookeeper connection client.
 *
 * @param {String} zkUrl, zookeeper server url
 * @param {Function} callBack, this is called after connection with zookeeper is established.
 */
function callZookeeper(zkUrl, callBack) {
    if (!_zooKeeperClient || (!!_zooKeeperClient.getState() && (
            _zooKeeperClient.getState() === zookeeper.State.EXPIRED ||
            _zooKeeperClient.getState() === zookeeper.State.AUTH_FAILED ||
            _zooKeeperClient.getState() === zookeeper.State.DISCONNECTED
        ))) {
        _zooKeeperClient = zookeeper.createClient(zkUrl);
        _zooKeeperClient.on('state', function (state) {
            if (state === zookeeper.State.EXPIRED || state === zookeeper.State.DISCONNECTED) {
                _zooKeeperClient = undefined;
            }
        });
        _zooKeeperClient.once("connected", function () {
            callBack(_zooKeeperClient);
        });
        _zooKeeperClient.connect();
    } else {
        callBack(_zooKeeperClient);
    }
}

/**
 * Watched event on the service node, which is triggered when the node is either deleted, modified, created or
 * children are changed.
 *
 * @param {Object} zkProps, Object containing zooKeeper server ip, port and path
 * @param {String} serviceName, Name of the service to be fetched
 * @param {Object} serviceClient, thrift client for the service
 * @param {async} async, to enable TFramed protocol
 * @param {Object} event, see @{zookeeper.Event}
 */
function zkWatcher(zkProps, serviceName, serviceClient, async, event) {
    if (zookeeper.Event.NODE_DELETED === event || zookeeper.Event.NODE_DATA_CHANGED === event
        || zookeeper.Event.NODE_CREATED === event || zookeeper.Event.NODE_CHILDREN_CHANGED === event) {
        zkHelper.createServiceClient(zkProps.zkUrl, serviceName, serviceClient, async, null);
    }
}

/**
 * Fetches children node for a service.
 * A cluster of pool is created for the service. For each child, a pool is added to the cluster.
 * Before fetching children details, if any previous cluster is drained, i.e. all the pools in the cluster
 * are drained gracefully.
 *
 * @param {Object} zkProps, Object containing zooKeeper server ip, port and path
 * @param {String} serviceName, Name of the service to be fetched
 * @param {Object} serviceClient, thrift client for the service
 * @param {Boolean} _async, to enable TFramed protocol
 * @param {Object} err, zookeeper connection error
 * @param {Array} children, all the running nodes of service
 */
function zkProcessChildren(zkProps, serviceName, serviceClient, _async, err, children) {
    var _cacheKey = _util.getServiceClientCacheKey(zkProps, serviceName);
    var _cluster = _clusterMap[_cacheKey];
    var zkClient = this;
    if (!!_cluster) {
        _cluster.clear();
        delete _clusterMap[_cacheKey];
    }
    if (!!err) {
        clearCbQueue(_cacheKey, undefined, err);
        return;
    }
    _cluster = _util.initCluster(serviceName, _config);
    var _nodeDetailsFn = [];
    children.forEach(function (node) {
        _nodeDetailsFn.push(_util.getNodeDetails.bind(zkClient, zkProps, serviceName, node, _cluster, _async));
    });
    async.parallel(_nodeDetailsFn, function () {
        var _serviceClientClone = serviceClientWrapper(serviceClient, _cluster);
        _serviceClients[_cacheKey] = _serviceClientClone;
        clearCbQueue(_cacheKey, _serviceClientClone, err);
    });
}

/**
 * Clears all the callbacks added to the queue.
 * Callback is invoked with arguments serviceClient and err object, if any.
 *
 * @param {String} cacheKey, cache key for call back array.
 * @param {Object} serviceClient, thrift service client
 * @param {Object} err, err, if any while connecting via zookeeper
 */
function clearCbQueue(cacheKey, serviceClient, err) {
    var cbQueue = _initCbQueue[cacheKey];
    while (cbQueue.length != 0) cbQueue.splice(0, 1)[0](serviceClient, err);
    delete _initCbQueue[cacheKey];
}

module.exports = zkHelper;

lib/helper/serviceClientWrapper.js
var __slice = [].slice;
var _ = require("underscore");
_.mixin(require("underscore.deep"));
var thrift = require('thrift');
var TIMEOUT_MESSAGE = "ZooKeeper-pool: Connection timeout";
var CLOSE_MESSAGE = "ZooKeeper-pool: Connection closed";

/**
 * Error handlers
 *
 * @param cb
 * @return {{error: Function, timeout: Function, close: Function}}
 */
function getErrorHandlers(cb){
    return {
        error: function(err){
            return cb(err);
        },
        timeout: function(){
            return cb(new Error(TIMEOUT_MESSAGE));
        },
        close: function(){
            return cb(new Error(CLOSE_MESSAGE));
        }
    };
}

/**
 * Binds listeners to the connection client.
 *
 * @param {Object} connection, thrift connection client
 * @param {Object} errorHandlers
 * @return {EventEmitter|*}
 */
function addListeners(connection, errorHandlers) {
    connection.on("error", errorHandlers.error);
    connection.on("close", errorHandlers.close);
    return connection.on("timeout", errorHandlers.timeout);
}

/**
 * Removes listeners added to connection client, after getting response from the client
 *
 * @param {Object} connection, thrift connection client
 * @param {Object} errorHandlers
 * @return {EventEmitter|*}
 */
function removeListeners(connection, errorHandlers) {
    connection.removeListener("error", errorHandlers.error);
    connection.removeListener("close", errorHandlers.close);
    return connection.removeListener("timeout", errorHandlers.timeout);
}

/**
 * Releases all pool resources acquired after processing the request.
 * It also removes all the added listeners to the connection client.
 * After releasing all the resources, it then calls the callback provided by the user.
 *
 * @param callback
 * @param errorHandlers, all the handlers(error, close & timeout) binded to connection.
 * @param pool, one of the generic pools from cluster
 * @param err, error object if any error occur while acquiring resource.
 * @param resp, resp from the server
 * @return {*}
 */
function releaseResources(callback, errorHandlers, pool, err, resp){
    var connection = this;
    removeListeners(connection, errorHandlers);
    pool.release(connection);
    return callback(err, resp);
}

/**
 * Wraps a function around the provided method in the thrift service client, as when
 * user calls this method, it first acquire the resource from the pool and then calls the original
 * service client method.
 * After getting response from the server, it releases the acquired resource back into the pool.
 *
 * @param {Object} cluster, cluster of pools for acquiring thrift connections from different servers.
 * @param {Function} fn, function reference to the provided method name.
 * @param {String} methodName, name of the function which is called.
 * @return {Function}, wrapped function with pool acquisition.
 */
function wrapperFn(cluster, fn, methodName) {
    var serviceClient = this;
    return function () {
        var args; //arguments to be passed to the service client method
        var cb; //callback that will be called after getting response from the client.

        //converting implicit argument as to pass it to client method and scraping callback
        if (arguments.length >= 2) {
            args = __slice.call(arguments, 0, arguments.length - 1);
            cb = arguments[arguments.length - 1];
        } else {
            args = [];
            cb = arguments[0];
        }
        if (!cluster) {
            return cb(new Error("Connection pooling: Not able to find any cluster"));
        }

        //acquiring pool from cluster with min waiting queue
        cluster.acquire(function (err, connection, pool) {
            if(!!err)
                return cb(err);

            var thriftClient;
            var errorHandlers = getErrorHandlers(cb);

            //adding error, timeout and close handlers on thrift connection
            addListeners(connection, errorHandlers);

            //creating thrift service client
            thriftClient = thrift.createClient(serviceClient, connection);

            //adding releaseResources as callback argument to thrift service client
            args.push(releaseResources.bind(connection, cb, errorHandlers, pool));

            //calling method in service client
            return thriftClient[methodName].apply(thriftClient, args);
        });
    }
}

/**
 * Service client wrapper.
 *
 * @param {Object} serviceClient, thrift client for the service
 * @param {Object} cluster, cluster of pools for acquiring thrift connections from different servers.
 * @return {Object}, cloned service client
 */
module.exports = function (serviceClient, cluster) {
    //creating a clone of client
    var serviceClientClone = _.clone(serviceClient.Client.prototype);

    /*wrapping client method with our wrapper function so a to initiate pooling when
     *method is invoked
     * */
    return _.mapValues(serviceClientClone, wrapperFn.bind(serviceClient, cluster));
};

lib/helper/util.js
var clusterPool = require('node-pool-cluster');
var _ = require('underscore');
var thrift = require('thrift');
var TIMEOUT_MESSAGE = "ZooKeeper-pool: Connection timeout";
var CLOSE_MESSAGE = "ZooKeeper-pool: Connection closed";
var zookeeper = require('node-zookeeper-client');

var serverSetHelper = {};

/**
 * Create a key for storing details the object in the cache.
 * Key is created as a combination of zookeeper url, path and service name
 *
 * @param {Object} zkProps, Object containing zooKeeper server ip, port and path
 * @param {string} serviceName, service name to be fetched
 * @return {string}, cache key
 */
serverSetHelper.getServiceClientCacheKey = function (zkProps, serviceName) {
    return zkProps.zkUrl + "~" + zkProps.zkPath + "~" + serviceName;
};


/**
 * Creates a cluster pool of a service.
 *
 * @param {String} clusterName, name of the cluster(service)
 * @param {Object} opts, pool factory options, similar to generic pool
 * @return {Object} returns cluster instance
 */
serverSetHelper.initCluster = function (clusterName, opts) {
    return clusterPool.initCluster({
        name: clusterName,
        destroy: function (connection) {
            return connection.end();
        },
        validate: function (connection) {
            return !connection.__ended;
        },
        max: opts.max || 3,
        min: opts.min || 0,
        refreshIdle: opts.refreshIdle || true,
        reapIntervalMillis: opts.reapIntervalMillis || 1000,
        idleTimeoutMillis: opts.idleTimeoutMillis || 20000
    }, true);
};


/**
 * Creates a connection to thrift service running on the specified host.
 *
 * @param {String} serviceName, Name of the service to be fetched
 * @param {Object} serverInfo, info regarding server, its ip and port
 * @param {async} async, to enable TFramed protocol
 * @param {Function} callBack, function which is called on successful connection or when error occurs.
 * @return {EventEmitter|*}
 */
serverSetHelper.createThriftConn = function (serviceName, serverInfo, async, callBack) {
    callBack = _.once(callBack);
    console.log("[Node-Conn-%s]%s:%d", serviceName, serverInfo.host, serverInfo.port);
    var connection = thrift.createConnection(serverInfo.host, serverInfo.port, {
        transport: (!!async ? thrift.TFramedTransport : thrift.TBufferedTransport),
        protocol: thrift.TBinaryProtocol
    });
    connection.__ended = false;
    connection.on("connect", function () {
        connection.connection.setKeepAlive(true);
        return callBack(null, connection);
    });
    connection.on("error", function (err) {
        connection.__ended = true;
        return callBack(err);
    });
    connection.on("close", function () {
        connection.__ended = true;
        return callBack(new Error(CLOSE_MESSAGE));
    });
    return connection.on("timeout", function () {
        connection.__ended = true;
        return callBack(new Error(TIMEOUT_MESSAGE));
    });
};

/**
 * It fetches details of all the running server instances(ip and port) of the specified service.
 *
 * @param {Object} zkProps, Object containing zooKeeper server ip, port and path
 * @param {String} serviceName, Name of the service to be fetched
 * @param {String} node, member node of the service containing information of server ip and port
 * @param {Object} cluster, cluster of pool for specified service
 * @param {async} async, to enable TFramed protocol
 * @param cb
 */
serverSetHelper.getNodeDetails = function(zkProps, serviceName, node, cluster, async, cb){
    var _pool, zkClient = this;
    zkClient.getData(zkProps.zkPath + "/nodeStatus/" + serviceName + "/" + node,
        function (event) {
            if (zookeeper.Event.NODE_DELETED === event || zookeeper.Event.NODE_DATA_CHANGED === event) {
                cluster.removePool(_pool);
                serverSetHelper.getNodeDetails().call(zkClient, zkProps, serviceName, node, cluster, async);
            }
        },
        function (err, data) {
            var _nodeDetails = JSON.parse(data.toString());
            var _host, _port;
            if (_nodeDetails.status == "ALIVE") {
                _host = _nodeDetails["serviceEndpoint"]["host"];
                _port = _nodeDetails["serviceEndpoint"]["port"];
                _pool = cluster.addPool(function (callBack) {
                    return serverSetHelper.createThriftConn(serviceName, {host: _host, port: _port}, async, callBack);
                });
            }
            !!cb && cb();
        }
    );
};

module.exports = serverSetHelper;